library/core/src/ffi/c_str.rs

```diff
@@ -169,24 +182,45 @@ impl fmt::Debug for CStr {
     }
 }
 
-#[stable(feature = "cstr_default", since = "1.10.0")]
-impl Default for &CStr {
-    fn default() -> Self {
-        const SLICE: &[c_char] = &[0];
-        // SAFETY: `SLICE` is indeed pointing to a valid nul-terminated string.
-        unsafe { CStr::from_ptr(SLICE.as_ptr()) }
+/// Converts a string literal to a `&'static Cstr`.
+///
+/// # Panics
+///
+/// `cstr!` panics if the input contains any interior nul bytes.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(const_cstr_from_bytes)]
+///
+/// use core::ffi::CStr;
+///
+/// const HELLO: &CStr = cstr!("Hello, world!");
+/// assert_eq!(HELLO.to_bytes_with_nul(), b"Hello, world!\0");
+/// ```
+#[macro_export]
+#[unstable(feature = "const_cstr_from_bytes", issue = "none")]
+#[rustc_diagnostic_item = "core_cstr_macro"]
+macro_rules! cstr {
+    ($($s:expr),*) => {
```

Comment:
Why does this accept `expr` instead of `literal`?

I would not expect this to convert identifiers and other expression syntax into a string. In fact, I'd expect it to fail if passed anything besides a byte or unicode string literal.

Reply:
Side note, I think we should be sure to accept byte strings since CStr allows non-UTF-8 bytes.

Reply:
Is there a way to differentiate between string and byte string literals without making this a proc-macro?

Reply:
I mean, an unstable trait seems more than sufficient. But a proc macro also wouldn't be too hard imho.

Reply:
I don't quite see how this can be solved with a trait, can you elaborate?

Reply:
Essentially, have a trait `CStrMacro` with a method `build` that's implemented only for byte slices and strings. You can make the trait const with an unstable feature and then just use the trait in the macro.

Reply:
It should be possible to accept a argument that can be referred to in compile time (e.g. a `const`) where concatenation is done through creating arrays and copying bytes instead of `concat!`.

Reply:
I mean I understand how to make a `const` trait and implement it for byte slices and strings, but e.g. a `build` method would need to be generic over input length `N` with output length `N + 1`, is that possible?

Reply:
>  a build method would need to be generic over input length N with output length N + 1, is that possible?

In stable Rust, it isn't, but there is the `generic_const_exprs` feature. The trait would have to be unstable anyways, so you probably can use that feature.

Reply:
Couldn't think of a way to do this with `generic_const_exprs` since I don't know how `&str` can be generic that way.

But managed to find a simpler way.